packages/backend/src/core/balances/BalanceProvider.ts

```diff
@@ -1,22 +1,219 @@
+import { getTokenByAssetId } from '@l2beat/config'
 import {
+  assert,
   AssetId,
+  Bytes,
   ChainId,
   EthereumAddress,
   UnixTime,
 } from '@l2beat/shared-pure'
+import { BigNumber, utils } from 'ethers'
+import { partition } from 'lodash'
 
 import { BalanceRecord } from '../../peripherals/database/BalanceRepository'
+import { EthereumClient } from '../../peripherals/ethereum/EthereumClient'
+import { MulticallClient } from '../../peripherals/ethereum/MulticallClient'
+import {
+  ARBITRUM_MULTICALL_ADDRESS,
+  ARBITRUM_MULTICALL_BLOCK,
+  ETHEREUM_MULTICALL_V1_ADDRESS,
+  ETHEREUM_MULTICALL_V1_BLOCK,
+} from '../../peripherals/ethereum/MulticallConfig'
+import {
+  MulticallRequest,
+  MulticallResponse,
+} from '../../peripherals/ethereum/types'
+
+const erc20Interface = new utils.Interface([
+  'function balanceOf(address account) view returns (uint256)',
+])
+
+const multicallInterface = new utils.Interface([
+  'function getEthBalance(address account) view returns (uint256)',
+])
 
 export interface BalanceQuery {
   holder: EthereumAddress
   assetId: AssetId
 }
 
-export interface BalanceProvider {
-  fetchBalances: (
-    missingAssets: BalanceQuery[],
+export interface NativeBalanceEncoding {
+  sinceBlock: number
+  address: EthereumAddress
+  encode: (address: EthereumAddress) => Bytes
+  decode: (response: Bytes) => bigint
+}
+
+export const ETHEREUM_BALANCE_ENCODING: NativeBalanceEncoding = {
+  sinceBlock: ETHEREUM_MULTICALL_V1_BLOCK,
+  address: ETHEREUM_MULTICALL_V1_ADDRESS,
+  encode: encodeGetEthBalance,
+  decode: decodeGetEthBalance,
+}
+
+export const ARBITRUM_BALANCE_ENCODING: NativeBalanceEncoding = {
+  sinceBlock: ARBITRUM_MULTICALL_BLOCK,
+  address: ARBITRUM_MULTICALL_ADDRESS,
+  encode: encodeGetEthBalance,
+  decode: decodeGetEthBalance,
+}
+
+export class BalanceProvider {
+  constructor(
+    private readonly ethereumClient: EthereumClient,
+    private readonly multicallClient: MulticallClient,
+    private readonly chainId: ChainId,
+    private readonly nativeBalanceEncoding: NativeBalanceEncoding | undefined,
```

Comment:
maybe this param can be stored in Multicall?

Reply:
nativeBalanceEncoding

Reply:
It's actually only partially related. While it's true that the multicall contract is the vehicle by which we query for eth balance it doesn't feel right to me that the MulticallClient is responsible for this encoding and decoding.

I thought about it but came to the conclusion that ultimately it is the responsibility of the caller to encode and decode calls and so calls to multicall itself should be no different